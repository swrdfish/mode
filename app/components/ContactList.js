import React from 'react'
import { connect } from 'react-redux'
import fuzzysearch from 'fuzzysearch'

import ContactProfilePicture from './ContactProfilePicture.js'
import { chat } from '../actions'


class ContactList extends React.Component {
    constructor(props){
        super(props)
        this.state = {
            selected: ""
        }
        this.handleClick = this.handleClick.bind(this)
    }

    handleClick(e) {
        let selected = e.currentTarget.getAttribute("data-id")
        if (selected) {
            this.setState({selected})
            this.props.dispatch(chat(selected))
            this.props.room.connectionManager.getConnection(selected)
        }
    }

    render() {  
        let userList = this.props.userList || []
        let userToDisplay = userList.map(user => ({
                "name": user.name? user.name : user.localIP,
                "id": user.uid,
                "unseen": user.unseen
            })
        )

        if(this.props.searchText !== ''){
            userToDisplay = userToDisplay.filter(contact => (fuzzysearch(this.props.searchText, contact.name.toLowerCase())))
        }

        let contacts = userToDisplay.map(contact => {
            let className = this.state.selected === contact.id ? 'contact-list-item selected' : 'contact-list-item'
            let unseen = contact.unseen?(
                <div className="unseen-count-wrapper">
                    <span className="unseen-count">{contact.unseen}</span>
                </div>
            ) : ''
            return (
                <div className={className} key={contact.id} data-id={contact.id} onClick={this.handleClick} >
                    <ContactProfilePicture profilePicture="/static/img/default_contact.jpg" />
                    <div className="contact-name">
                        {contact.name}
                    </div>
                    {unseen}
                </div>
            );
        }, this);

        return (
            <div>
                {(contacts.length > 0)? contacts : (<div className="contact-not-found">No contacts found.</div>)}
            </div>
        )
    }
}


const mapStateToProps = ({userList, contactFilter, room, dispatch}) => ({
    searchText: contactFilter,
    userList,
    room,
    dispatch
})

let ConnectedContactList = connect(mapStateToProps)(ContactList) 
export default ConnectedContactList