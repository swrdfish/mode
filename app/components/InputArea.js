import React from 'react'


class InputArea extends React.Component {
	render() {
		return (
			<div className="input-area-wrapper">
				<input type="text" placeholder="Type message" className="input-area-inputfiled" />
				<i className="fa fa-paperclip" id="attach-button" aria-hidden="true"></i>
			</div>
		)
	}
}

export default InputArea