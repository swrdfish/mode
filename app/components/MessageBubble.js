import React from 'react'
import AutoLinker from 'autolinker'
import escapeHTML from 'escape-html'


class MessageBubble extends React.Component {
    constructor(props) {
        super(props)
    }

    render() {
        let item = this.props.item
        let className = item.isMine?'message-bubble mine':'message-bubble'
        let downloadButtonClass = item.isMine?"":"fa fa-download"
        let pretext = item.isMine?"Sent ":"Download "
        let text

        let itemRendered
        if(item.type == "file") {
            if(item.body.count > 1) {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.count} files&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            } else {
                text = (
                    <a href="#" onClick={this.props.onDownload} data-id={item.body.id}>
                        {item.body.filenames[0]}&nbsp;&nbsp;&nbsp;
                        <span className={downloadButtonClass + " download-button"}/>
                    </a>
                )
            }

            if(item.isMine) {
                text = "" + item.body.count
                text += item.body.count > 1?" files":" file"
            }
            className += " file"
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span>
                            {pretext}&nbsp;&nbsp;{text}               
                        </span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        } else {
            let messageBody = AutoLinker.link(escapeHTML(item.body), {className: "message-bubble-link"})
            itemRendered = (
                <div className='message-bubble-wrapper'>
                    <div className={className}>
                        <span className='message-bubble-body' dangerouslySetInnerHTML={{ __html: messageBody }}></span>
                        <span className='time-stamp'>{item.timeStamp.toLocaleString('en-US', { hour: 'numeric', minute: 'numeric', hour12: true })}</span>
                    </div>
                </div>
            )
        }
        return itemRendered
    }
}

export default MessageBubble