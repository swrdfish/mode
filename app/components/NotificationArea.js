import React from 'react'
import { connect } from 'react-redux'
import {unnotify} from '../actions'


class NotificationArea extends React.Component {
  constructor(props){
    super(props)
    this.handleClearNotification = this.handleClearNotification.bind(this)
  }

  handleClearNotification(event) {
    this.props.dispatch(unnotify(event.target.getAttribute('data-notification-idx')))
  }

  render() {
    let displayNotifications = this.props.notifications.map((notify, index) => {
      return (
        <div className={notify.type == 'error'? 'notification error': 'notification'} key={index}>
          {notify.msg}
          <i className="fa fa-close clear-notification" aria-hidden="true" data-notification-idx={index} onClick={this.handleClearNotification}></i>
        </div>
      )
    })

    return (
      <div className="notification-area">
        {displayNotifications}
      </div>
    )
  }
}

const mapStateToProps = ({notifications, dispatch}) => ({
  notifications,
  dispatch
})

let ConnectedNotificationArea = connect(mapStateToProps)(NotificationArea) 
export default ConnectedNotificationArea;