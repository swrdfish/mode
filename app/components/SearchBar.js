import React from 'react'
import { connect } from 'react-redux'

import { setContactFilter } from '../actions'


export class SearchBar extends React.Component{
	constructor(props) {
		super(props)
		this.state = {value: ''}
		this.handleOnChange = this.handleOnChange.bind(this)
		this.handleClearSearch = this.handleClearSearch.bind(this)
		this.handleKeyDown = this.handleKeyDown.bind(this)
		this.closeButton = (<i className="fa fa-close clear-search" aria-hidden="true" onClick={this.handleClearSearch}></i>)
		this.searchButton = (<i className="fa fa-search" aria-hidden="true"></i>)
	}

	handleOnChange(event) {
		this.setState({value: event.target.value})
		this.props.dispatch(setContactFilter(event.target.value.toLowerCase()))
	}

	handleClearSearch(event) {
		this.setState({value: ''})
		this.props.dispatch(setContactFilter(''))
	}

	handleKeyDown(event) {
		if(event.key === "Escape"){
			event.stopPropagation()
			event.preventDefault()
			this.setState({value: ''})
			this.props.dispatch(setContactFilter(''))
		}
	}

	render() {
		return (
			<div className="search-wrapper">
				<input
					placeholder="search users"
					value={this.state.value}
					onKeyDown={this.handleKeyDown}
					onChange={this.handleOnChange}
					type="text"
				/>
				{this.state.value != '' ? this.closeButton : this.searchButton}
			</div>
		)
	}
}


let ConnectedSearchBar = connect()(SearchBar) 
export default ConnectedSearchBar;
