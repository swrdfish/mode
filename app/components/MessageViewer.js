import React from 'react'
import MessageBubble from "./MessageBubble"


class MessageViewer extends React.Component {
    render(){
        if(this.props.chatter) {
            let renderedMsg = (<div className='empty-message-view'>Say something!</div>)
            if (this.props.messages) {
                let messages = this.props.messages.reduceRight((arr, item) => {
                    return arr.concat((<MessageBubble item={item} key={item.id} onDownload={this.props.onDownload}/>))
                }, [])
                renderedMsg = (
                    <div className='messages-scrollable'>
                        {messages}
                    </div>
                )
            }

            return (
                <div className='message-view-wrapper'>
                    {renderedMsg}
                </div>
            )
        } else {
            return (
                <div className='message-view-wrapper'>
                    <div className='empty-message-view'>select someone to chat with.</div>
                </div>
            )
        }
    }
}

export default MessageViewer