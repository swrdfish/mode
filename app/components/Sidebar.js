import React from 'react'

import UserInfoDisplay from './UserInfoDisplay'
import SearchBar from './SearchBar'
import ContactList from './ContactList'


class Sidebar extends React.Component {
    render() {
        return (
            <div id="sidebar" className="col-md-4">
                <UserInfoDisplay />
                <SearchBar />
                <ContactList />
            </div>
        )
    }

}

export default Sidebar