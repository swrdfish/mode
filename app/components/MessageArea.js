import React from 'react'
import { connect } from 'react-redux'
import uuid from 'node-uuid';

import InputArea from './InputArea.js'
import MessageViewer from './MessageViewer.js'
import {addMessage, addFiles} from '../actions'


export class MessageArea extends React.Component {
    constructor(props) {
        super(props)
        this.onSendMessage = this.onSendMessage.bind(this)
        this.onSendFile = this.onSendFile.bind(this)
        this.onDownload = this.onDownload.bind(this)
        this.files = {}
    }

    onSendMessage(msg) {
        if(this.props.chatArea.uid && msg) {
            let connection = this.props.room.connectionManager.getConnection(this.props.chatArea.uid)
            if(connection.dataChannel.readyState == "open") {
                this.props.dispatch(addMessage(this.props.chatArea.uid, "text", msg, true))
                connection.dataChannel.send(JSON.stringify({
                    type: "text",
                    message: msg
                }))
            }
        }
    }

    onSendFile(files) {
        if(this.props.chatArea.uid && files) {
            let id = uuid.v4()
            this.props.dispatch(addFiles(id, files))
            let filenames = []
            for( let iii = 0; iii < files.length; iii++ ) {
                filenames.push(files.item(iii).name)
            }
            let msg = {
                count : files.length,
                id,
                filenames 
            }
            let connection = this.props.room.connectionManager.getConnection(this.props.chatArea.uid)
            if(connection.dataChannel.readyState == "open") {
                this.props.dispatch(addMessage(this.props.chatArea.uid, "file", msg, true))
                connection.dataChannel.send(JSON.stringify({
                    type: "file",
                    message: msg
                }))
            }
        }
    }

    onDownload(event) {
        let fileid = event.target.getAttribute("data-id")
        let connection = this.props.room.connectionManager.getConnection(this.props.chatArea.uid)
        connection.dataChannel.send(JSON.stringify({
            type: "get",
            message: fileid
        }))
        console.log("download the file", fileid)
    }

    render() {
        let chatter = this.props.chatArea.uid
        let messages = this.props.messages[chatter]
        let inputArea = (
            <InputArea 
                onSendMessage={this.onSendMessage}
                onSendFile={this.onSendFile} 
                onDownload={this.onDownload} />
        )

        return (
            <div id="message-area" className="col-md-8">
                <MessageViewer
                    chatter={chatter}
                    messages={messages}
                    onDownload={this.onDownload} />
                {chatter?inputArea:''}        
            </div>
        )
    }
}

let mapStateToProps = ({messages, chatArea, room, dispatch}) => ({
    messages,
    chatArea,
    room,
    dispatch
})

const ConnectedMessageArea = connect(mapStateToProps)(MessageArea)
export default ConnectedMessageArea