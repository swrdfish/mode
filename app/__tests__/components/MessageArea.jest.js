import React from 'react'
import { shallow } from 'enzyme'
import { MessageArea } from '../../components/MessageArea'
import renderer from 'react-test-renderer';


Date.now = jest.fn(() => new Date(Date.UTC(2017, 7, 9, 8)).valueOf())

describe('Component: MessageArea', () => {
    it('should render the profile picture', () => {
        let chatArea = {uid: 123456}
        let messages = {
            '123456': [{
                type: 'text',
                timeStamp: Date.now(),
                id: 123
            }] 
        }
        const wrapper = shallow(<MessageArea chatArea={chatArea} messages={messages}/>);
        expect(wrapper.find("#message-area")).toHaveLength(1)
    })

    it('should match previous snapshot', () => {
        let chatArea = {uid: 123456}
        let messages = {
            '123456': [{
                type: 'text',
                timeStamp: Date.now(),
                id: 123
            }] 
        }
        const tree = renderer
            .create(<MessageArea chatArea={chatArea} messages={messages}/>).toJSON();
        expect(tree).toMatchSnapshot();
    })
})