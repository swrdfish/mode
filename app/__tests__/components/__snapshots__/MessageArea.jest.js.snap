// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MessageArea should match previous snapshot 1`] = `
<div
  className="col-md-8"
  id="message-area"
>
  <div
    className="message-view-wrapper"
  >
    <div
      className="messages-scrollable"
    >
      <div
        className="message-bubble-wrapper"
      >
        <div
          className="message-bubble"
        >
          <span
            className="message-bubble-body"
            dangerouslySetInnerHTML={
              Object {
                "__html": "undefined",
              }
            }
          />
          <span
            className="time-stamp"
          >
            1,502,265,600,000
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="input-area-wrapper"
  >
    <input
      className="input-area-inputfiled"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Type message"
      type="text"
      value=""
    />
    <div
      className="file-input-wrapper fa fa-paperclip"
    >
      <input
        aria-hidden="true"
        className="file-input"
        id="attach-button"
        multiple="multiple"
        name="file-input"
        onChange={[Function]}
        type="file"
      />
    </div>
  </div>
</div>
`;
